File: 2019/day21/day21.py
#!/usr/bin/env python3
# https://adventofcode.com/2019/day/21

from pathlib import Path
import sys
from copy import deepcopy
from collections import defaultdict, deque
import re

filename = ("test.txt" if sys.argv[1] == "-t" else sys.argv[1]) if len(sys.argv) > 1 else "input.txt"
data = Path(filename).read_text()
lines = data.splitlines()


sys.path.append(Path(__file__).parent.parent.as_posix())
from intcode.Intcode import Computer


filename = sys.argv[1] if len(sys.argv) > 1 else "input.txt"
software = Path(filename).read_text()

droid = Computer()
droid.load(software)


def run_script(script):
    droid.flush_io()
    droid.input.extend(map(ord, script))
    droid.run()

    res = droid.output[-1]
    if res == 10:
        print("error")
    else:
        print(res)


# J = not A or not (C and D)
run_script(
    """\
NOT A J
NOT C T
AND D T
OR T J
WALK
"""
)


# J = not A or not (C and D)
run_script(
    """\
NOT B J
NOT C T
OR T J
AND D J
AND H J
NOT A T
OR T J
RUN
"""
)

File: 2019/day21/day21.rs
//! [Day 21: Springdroid Adventure](https://adventofcode.com/2019/day/21)

use intcode::{Computer, State};

struct Puzzle {
    springdroid: Computer,
}

impl Puzzle {
    /// Initialize from the puzzle input.
    fn new(data: &str) -> Self {
        Self {
            springdroid: Computer::load(data),
        }
    }

    fn run_script(&self, script: &str) -> i64 {
        let mut springdroid = self.springdroid.clone();

        springdroid.push_ascii(script);

        let mut robot_report = 0;
        loop {
            match springdroid.run() {
                State::Output(num) => {
                    // print!("{}", (num as u8) as char);
                    robot_report = num;
                }
                State::Input => panic!("missing input ?!"),
                State::Halted => break robot_report,
            };
        }
    }

    /// Solve part one.
    fn part1(&self) -> i64 {
        self.run_script(concat!(
            "NOT A J\n",
            "NOT C T\n",
            "AND D T\n",
            "OR T J\n",
            "WALK\n"
        ))
    }

    /// Solve part two.
    fn part2(&self) -> i64 {
        self.run_script(concat!(
            "NOT B J\n",
            "NOT C T\n",
            "OR T J\n",
            "AND D J\n",
            "AND H J\n",
            "NOT A T\n",
            "OR T J\n",
            "RUN\n"
        ))
    }
}

/// # Panics
/// over malformed input
#[must_use]
pub fn solve(data: &str) -> (i64, i64) {
    let puzzle = Puzzle::new(data);
    (puzzle.part1(), puzzle.part2())
}

pub fn main() {
    let args = aoc::parse_args();
    args.run(solve);
}
